code/modules/clothing/clothing.dm

```diff
@@ -398,11 +398,158 @@
 	var/vision_flags = 0
 	var/see_in_dark = 0
 	var/lighting_alpha
+	/// the head clothing that this item may be attached to.
+	var/obj/item/clothing/under/has_under = null
+	/// the attached hats to this hat.
+	var/list/attached_hats = list()
+	/// if this hat can have hats placed ontop of it.
+	var/can_have_hats = FALSE
+	/// if this hat can be a hat of a hat. Hat^2
+	var/can_be_hat = TRUE
 
 /obj/item/clothing/head/update_icon_state()
 	if(..())
 		item_state = "[replacetext("[item_state]", "_up", "")][up ? "_up" : ""]"
 
+/obj/item/clothing/head/AltShiftClick(mob/user)
+	if(user.stat || HAS_TRAIT(user, TRAIT_HANDS_BLOCKED) || !Adjacent(user))
+		return
+	if(!length(attached_hats))
+		return
+	var/obj/item/clothing/head/hat
+	if(length(attached_hats) > 1)
+		var/pick = radial_menu_helper(usr, src, attached_hats, custom_check = FALSE, require_near = TRUE)
+		if(!pick)
+			return
+		hat = pick
+	else
+		hat = attached_hats[1]
+	remove_hat(user, hat)
+
+/obj/item/clothing/head/Destroy()
+	if(is_equipped())
+		var/mob/M = loc
+		for(var/obj/item/clothing/head/H as anything in attached_hats)
+			H.on_removed(M)
+	else
+		for(var/obj/item/clothing/head/H as anything in attached_hats)
+			H.forceMove(loc)
+	attached_hats = null
+	return ..()
+
+/obj/item/clothing/head/dropped(mob/user, silent)
+	..()
+	if(!ishuman(user))
+		return
+	var/mob/living/carbon/human/H = user
+	if(H.get_item_by_slot(SLOT_HUD_HEAD) == src)
+		for(var/obj/item/clothing/head/hat as anything in attached_hats)
+			hat.attached_unequip()
+
+/obj/item/clothing/head/examine(mob/user)
+	. = ..()
+	if(length(attached_hats))
```

Comment:
Shouldn't need this length check, you should be able to just try to iterate over it which'll do nothing if it's empty